### PR TITLE
0.3

### LINKED ISSUE #168: Conversion to `NonEmpty` rather than list for types which should do it naturally
Could we get `Foldable1` on types which can implement it.
For instance for `Block` it would be nice to be able to convert it to `NonEmpty Statement` rather than `[Statement]`. I have very little idea about lenses but I understand that in order to use `toNonEmptyOf` the type should implement `Foldable1` and/or `Traversable1` rather then `Foldable` as it does now. 

The same I guess applies to `CommaSep1` and `CommaSep1'`

### PR DESCRIPTION
**Pull request by LightAndLight**
Fixes #168. See changelog for details.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -16,102 +18,101 @@ cache:
...
-  - HADDOCK=${HADDOCK-true}
-  - UNCONSTRAINED=${UNCONSTRAINED-true}
-  - NOINSTALLEDCONSTRAINTS=${NOINSTALLEDCONSTRAINTS-false}
+  - TEST=--enable-tests
+  - BENCH=--enable-benchmarks
   - GHCHEAD=${GHCHEAD-false}
-  - travis_retry cabal update -v
-  - "sed -i.bak 's/^jobs:/-- jobs:/' ${HOME}/.cabal/config"
+  - travis_retry ${CABAL} update -v
+  - sed -i.bak 's/^jobs:/-- jobs:/' $CABALHOME/config
```
Comment (endgame) [inline .travis.yml]: You can use `&` to mean "the entire match":
```
$ echo "jobs: 4" | sed 's/^jobs:/-- &/'
-- jobs: 4
```

File: .travis.yml [hunk 2]
```diff
@@ -16,102 +18,101 @@ cache:
...
   - rm -fv cabal.project cabal.project.local
-  - grep -Ev -- '^\s*--' ${HOME}/.cabal/config | grep -Ev '^\s*$'
-  - "printf 'packages: \".\" \"./hpython-test\"\\n' > cabal.project"
+  - grep -Ev -- '^\s*--' $CABALHOME/config | grep -Ev '^\s*$'
+  - rm -f cabal.project
+  - touch cabal.project
+  - "printf 'packages: \".\"\\n' >> cabal.project"
```
Comment (endgame) [inline .travis.yml]: `echo -e` I think is more idiomatic than `printf`, and do you need the outer quotes? if not, you can unescape the inner `"` ones.

File: .travis.yml [hunk 3]
```diff
@@ -16,102 +18,101 @@ cache:
...
+  - "printf 'packages: \"./hpython-test\"\\n' >> cabal.project"
+  - "printf 'write-ghc-environment-files: always\\n' >> cabal.project"
   - touch cabal.project.local
-  - "if ! $NOINSTALLEDCONSTRAINTS; then for pkg in $($HCPKG list --simple-output); do echo $pkg  | grep -vw -- hpython | grep -vw -- hpython-test | sed 's/^/constraints: /' | sed 's/-[^-]*$/ installed/' >> cabal.project.local; done; fi"
+  - "for pkg in $($HCPKG list --simple-output); do echo $pkg | sed 's/-[^-]*$//' | grep -vE -- '^(hpython|hpython-test)$' | sed 's/^/constraints: /' | sed 's/$/ installed/' >> cabal.project.local; done"
```
Comment (endgame) [inline .travis.yml]: Worth making a standalone shell script?

File: .travis.yml (hunk 1 above)
Comment (LightAndLight) [reply .travis.yml]: Thanks, but I think this is an auto-generated file

File: src/Data/Validate/Monadic.hs [hunk 4]
```diff
@@ -38,9 +38,14 @@ newtype ValidateM e m a = ValidateM { unValidateM :: Compose m (Validation e) a
 runValidateM :: ValidateM e m a -> m (Validation e a)
 runValidateM = getCompose . unValidateM
 
--- | Bind into a 'ValidateM'. Note that the first parameter is @m a@, not @ValidateM e m a@.
-bindVM :: Monad m => m a -> (a -> ValidateM e m b) -> ValidateM e m b
-bindVM m f = ValidateM . Compose $ m >>= getCompose . unValidateM . f
+-- | Bind into a 'ValidateM'
+bindVM :: Monad m => ValidateM e m a -> (a -> ValidateM e m b) -> ValidateM e m b
```
Comment (endgame) [inline src/Data/Validate/Monadic.hs]: Is there an `instance Monad m => Monad (ValidateM e m)`? If so, do you need this?

File: src/Language/Python/Validate/Scope.hs [hunk 5]
```diff
@@ -1,10 +1,11 @@
 {-# language DataKinds, TypeOperators #-}
-{-# language GeneralizedNewtypeDeriving #-}
-{-# language TemplateHaskell, TypeFamilies, FlexibleInstances, MultiParamTypeClasses #-}
+{-# language DeriveFunctor #-}
 {-# language FlexibleContexts #-}
 {-# language RankNTypes #-}
 {-# language LambdaCase #-}
+{-# language OverloadedLists #-}
 {-# language ScopedTypeVariables, TypeApplications #-}
+{-# language MultiParamTypeClasses #-}
```
Comment (endgame) [inline src/Language/Python/Validate/Scope.hs]: How are you deciding whether to list extensions on multiple lines or not?

File: src/Data/Validate/Monadic.hs (hunk 4 above)
Comment (LightAndLight) [reply src/Data/Validate/Monadic.hs]: No, it's not a monad, for the same reason that `Validation` isn't

File: src/Language/Python/Validate/Scope.hs (hunk 5 above)
Comment (LightAndLight) [reply src/Language/Python/Validate/Scope.hs]: They're loosely grouped by 'topic'. So scoped type variables and type applications are like 'extra system F annotations stuff'.

**Pull request merged**